src/python/pants/backend/codegen/protobuf/lint/buf/format_rules.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from dataclasses import dataclass
+
+from pants.backend.codegen.protobuf.lint.buf.skip_field import SkipBufField
+from pants.backend.codegen.protobuf.lint.buf.subsystem import BufSubsystem
+from pants.backend.codegen.protobuf.target_types import (
+    ProtobufDependenciesField,
+    ProtobufSourceField,
+)
+from pants.core.goals.fmt import FmtRequest, FmtResult
+from pants.core.goals.lint import LintResult, LintResults, LintTargetsRequest
+from pants.core.util_rules.external_tool import DownloadedExternalTool, ExternalToolRequest
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.core.util_rules.system_binaries import (
+    SEARCH_PATHS,
+    BinaryPath,
+    BinaryPathRequest,
+    BinaryPaths,
+    BinaryShims,
+    BinaryShimsRequest,
+)
+from pants.engine.fs import Digest, MergeDigests
+from pants.engine.internals.native_engine import Snapshot
+from pants.engine.platform import Platform
+from pants.engine.process import FallibleProcessResult, Process, ProcessResult
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class BufFieldSet(FieldSet):
+    required_fields = (ProtobufSourceField,)
+
+    sources: ProtobufSourceField
+    dependencies: ProtobufDependenciesField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipBufField).value
+
+
+class BufRequest(LintTargetsRequest, FmtRequest):
+    field_set_type = BufFieldSet
+    name = BufSubsystem.options_scope
```

Comment:
Huh! We before did not have any instances of implementing the same `name` twice, which I think will mess up the final summary of `./pants lint` and mess up `--lint-only=buf`. I thought about eagerly validating that two `LintTargetsRequest` don't use the same name but didn't out of laziness. 

I think we need to use unique names, like `buf-fmt` and `buf-lint`. What names would make sense?

> I thought about eagerly validating that two `LintTargetsRequest` don't use the same name but didn't out of laziness. 

Our code to check for `generate-lockfiles` below. If you had any interest, a PR would definitely be welcomed to add this check - I honestly didn't think we'd hit the edge case in the wild.

https://github.com/pantsbuild/pants/blob/0380b5578f9d5b4142ad6294d2eecaacd5b187b2/src/python/pants/core/goals/generate_lockfiles.py#L191-L208

--

I think we probably should update 2.11 before it's stabilized to use whatever new name you recommend like `buf-lint`. Otherwise it would break our deprecation policy because `--lint-only=buf` would no longer work in 2.12 but does in 2.11. What do you think about cherry-picking this new feature overall to 2.11? It's acceptable according to our release policy because this is lower risk + it thematically fits with a 2.11 feature we already had.

Reply:
I was getting a bit confused myself with the options scopes. 😅 In the current 2.11 Buf stuff, the buf subsystem (and thus the linter) has the options scope `buf-lint`. I've not changed `BufSubsystem.options_scope` to `buf`, `BufLintRequest.name` to `buf lint` and `BufFormatRequest.name` to `buf-format` - but I'll admit I'm not entirely sure how this will behave.

Which options/flags/etc. are based on request name vs subsystem options scope? I know the rules themselves check `BufSubsystem.skip`, so that would be `--buf-skip`? Is `--lint-only` on a request basis, e.g. `--lint-only=buf-format`? 

Regardless, I'm fine with cherry-picking this to 2.11 to reduce the potential deprecation mess in future versions. Just need to figure out these options, but my knowledge about that part of Pants is still quite limited.

Reply:
But having the subsystem named a generic `buf` makes it impossible to send separate passthrough args for each request. So should `lint_rules.py` and `format_rules.py` have their own `Subsystem` instead that extends `BufSubsystem` and just changes the options scope?

Reply:
Oof that's awkward. I for sure think we should stick with `Buf` as an `ExternalTool` and only have that once; seems wrong (and verbose) to use a different version of Buf for formatting vs linting.

So then to handle the desire for different passthrough args & skip options, two options:

1. Add `buf-fmt` and `buf-lint` as additional subsystems, which augment `buf`
2. Stick with `buf`, and have dedicated options like `--fmt-args` and `--lint-args`. There's no need for us to stick with `--args`. Feel free to patch `ArgsListOption` to allow you to override the flag name like that.

I think I recommend 2 as less verbose? Not sure if we'd also want `--fmt-skip` and `--lint-skip`, vs a generic `--skip`?

Re `--lint-only`, that is solely based off of the `name` you set in `LintTargetsRequest`. So if you only have a single subsystem `buf`, it is fine to still set the names to `buf-fmt` and `buf-lint` or whatever you want.

Wdyt?

Reply:
I went with option 2 and with separate skips. Should be good to go now 👍